- Due Thursday 11/12 no later than 9:00 a.m., send email to datascience@codeup.com
- Submit link to GitHub notebook that asks and answers questions - document the work you do to justify findings
- Compose an email with the answers to the questions/your findings, and in the email, include the link to your notebook in GitHub and attach your slide. 
- You will not present this, so be sure that the details you need your need your leader to convey/understand are clearly communicated in the email. 
- Slide should be like an exec. Summary and be in form to present. 
- Continue using best practices of acquire.py, prepare.py, etc. 
- No modeling to be done, and no need to split the data into train/validate/test. 
- alumni.codeup.com has info about cohorts/dates/names

- 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?
- 2. Is there a cohort that referred to a lesson significantly more that other cohorts seemed to gloss over? 
- 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students? 
- 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses? Any odd user-agents? 
- 5. At some point in the last year, ability for students and alumni to cross-access curriculum (web dev to ds, ds to web dev) should have been shut off. Do you see any evidence of that happening? Did it happen before? 
- 6. What topics are grads continuing to reference after graduation and into their jobs (for each program)? 
- 7. Which lessons are least accessed? 
- 8. Anything else I should be aware of? 

In [2]:
df = pd.read_csv('curriculum.txt',
                engine='python',
                 header=None,
                 index_col=False,
                 sep=r'\s(?=(?:[^"]*"[^"]*")*[^"]*$)(?![^\[]*\])',
                 na_values='"-"',)

In [3]:
df.head()

0         1                                    2  3     4  \
0  2018-01-26  09:55:03                                    /  1   8.0   
1  2018-01-26  09:56:02                              java-ii  1   8.0   
2  2018-01-26  09:56:05  java-ii/object-oriented-programming  1   8.0   
3  2018-01-26  09:56:06   slides/object_oriented_programming  1   8.0   
4  2018-01-26  09:56:24            javascript-i/conditionals  2  22.0   

              5  
0  97.105.19.61  
1  97.105.19.61  
2  97.105.19.61  
3  97.105.19.61  
4  97.105.19.61

In [4]:
df.columns = ['date', 'time', 'page_viewed', 'user_id', 'cohort_id', 'ip']

In [5]:
df.head()

date      time                          page_viewed  user_id  \
0  2018-01-26  09:55:03                                    /        1   
1  2018-01-26  09:56:02                              java-ii        1   
2  2018-01-26  09:56:05  java-ii/object-oriented-programming        1   
3  2018-01-26  09:56:06   slides/object_oriented_programming        1   
4  2018-01-26  09:56:24            javascript-i/conditionals        2   

   cohort_id            ip  
0        8.0  97.105.19.61  
1        8.0  97.105.19.61  
2        8.0  97.105.19.61  
3        8.0  97.105.19.61  
4       22.0  97.105.19.61

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 719459 entries, 0 to 719458
Data columns (total 6 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   date         719459 non-null  object 
 1   time         719459 non-null  object 
 2   page_viewed  719458 non-null  object 
 3   user_id      719459 non-null  int64  
 4   cohort_id    674619 non-null  float64
 5   ip           719459 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 32.9+ MB


In [7]:
no_cohort_df = df[df['cohort_id'].isnull()]

In [8]:
no_cohort_df.head()

date      time                            page_viewed  user_id  \
411  2018-01-26  16:46:16                                      /       48   
412  2018-01-26  16:46:24  spring/extra-features/form-validation       48   
425  2018-01-26  17:54:24                                      /       48   
435  2018-01-26  18:32:03                                      /       48   
436  2018-01-26  18:32:17              mysql/relationships/joins       48   

     cohort_id            ip  
411        NaN  97.105.19.61  
412        NaN  97.105.19.61  
425        NaN  97.105.19.61  
435        NaN  97.105.19.61  
436        NaN  97.105.19.61

**Drop rows with null values and turn**

In [9]:
df.dropna(inplace=True)
# df.cohort_id = df.cohort_id.astype('int')
# df['date'] = df.date + " " + df.time
# df.drop(columns=('time'), inplace=True)
# df.date = pd.to_datetime(df.date)
# df = df.set_index('date')

In [10]:
df.head()

date      time                          page_viewed  user_id  \
0  2018-01-26  09:55:03                                    /        1   
1  2018-01-26  09:56:02                              java-ii        1   
2  2018-01-26  09:56:05  java-ii/object-oriented-programming        1   
3  2018-01-26  09:56:06   slides/object_oriented_programming        1   
4  2018-01-26  09:56:24            javascript-i/conditionals        2   

   cohort_id            ip  
0        8.0  97.105.19.61  
1        8.0  97.105.19.61  
2        8.0  97.105.19.61  
3        8.0  97.105.19.61  
4       22.0  97.105.19.61

In [11]:
cohort = pd.read_csv('cohort_name.csv')

In [12]:
cohort.head()

Unnamed: 0 Unnamed: 1 Unnamed: 2  Unnamed: 3  Unnamed: 4  Unnamed: 5  \
0         NaN  cohort_id       name  start_date    end_date  program_id   
1         NaN          1     Arches  2014-02-04  2014-04-22           1   
2         NaN          2   Badlands  2014-06-04  2014-08-22           1   
3         NaN          3   Carlsbad  2014-09-04  2014-11-05           1   
4         NaN          4     Denali  2014-10-20  2015-01-18           1   

   Unnamed: 6  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [14]:
# cohort[cohort.name == 'Ada']

In [15]:
cohort.columns = cohort.iloc[0]

In [16]:
cohort.head()

0  NaN  cohort_id      name  start_date    end_date  program_id  NaN
0  NaN  cohort_id      name  start_date    end_date  program_id  NaN
1  NaN          1    Arches  2014-02-04  2014-04-22           1  NaN
2  NaN          2  Badlands  2014-06-04  2014-08-22           1  NaN
3  NaN          3  Carlsbad  2014-09-04  2014-11-05           1  NaN
4  NaN          4    Denali  2014-10-20  2015-01-18           1  NaN

In [17]:
cohort = cohort.iloc[1:]

In [18]:
cohort.head()

0  NaN cohort_id        name  start_date    end_date program_id  NaN
1  NaN         1      Arches  2014-02-04  2014-04-22          1  NaN
2  NaN         2    Badlands  2014-06-04  2014-08-22          1  NaN
3  NaN         3    Carlsbad  2014-09-04  2014-11-05          1  NaN
4  NaN         4      Denali  2014-10-20  2015-01-18          1  NaN
5  NaN         5  Everglades  2014-11-18  2015-02-24          1  NaN

In [19]:
cohort = cohort[['cohort_id', 'name', 'start_date', 'end_date']]

In [20]:
cohort.head()

0 cohort_id        name  start_date    end_date
1         1      Arches  2014-02-04  2014-04-22
2         2    Badlands  2014-06-04  2014-08-22
3         3    Carlsbad  2014-09-04  2014-11-05
4         4      Denali  2014-10-20  2015-01-18
5         5  Everglades  2014-11-18  2015-02-24

In [21]:
cohort.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46 entries, 1 to 46
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   cohort_id   46 non-null     object
 1   name        46 non-null     object
 2   start_date  46 non-null     object
 3   end_date    46 non-null     object
dtypes: object(4)
memory usage: 1.6+ KB


In [22]:
cohort.cohort_id = cohort.cohort_id.astype('int')


In [23]:
df.cohort_id.value_counts()

28.0    60315
33.0    40168
29.0    37548
53.0    36047
24.0    35624
57.0    32447
56.0    31670
22.0    30328
51.0    29688
58.0    28354
32.0    28333
23.0    28329
26.0    27637
52.0    27518
25.0    25427
31.0    25253
34.0    25181
59.0    22425
27.0    20447
55.0    20410
61.0    11774
14.0     9495
1.0      8884
62.0     8718
21.0     7444
17.0     4925
13.0     2733
18.0     2005
8.0      1712
19.0     1165
16.0      743
15.0      691
7.0       495
12.0      302
11.0      208
2.0        93
6.0        72
9.0         5
4.0         4
5.0         1
Name: cohort_id, dtype: int64

In [24]:
result = pd.merge(df, cohort, on='cohort_id')

In [25]:
result.head()

date      time                          page_viewed  user_id  \
0  2018-01-26  09:55:03                                    /        1   
1  2018-01-26  09:56:02                              java-ii        1   
2  2018-01-26  09:56:05  java-ii/object-oriented-programming        1   
3  2018-01-26  09:56:06   slides/object_oriented_programming        1   
4  2018-01-26  10:40:15               javascript-i/functions        1   

   cohort_id            ip     name  start_date    end_date  
0        8.0  97.105.19.61  Hampton  2015-09-22  2016-02-06  
1        8.0  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2        8.0  97.105.19.61  Hampton  2015-09-22  2016-02-06  
3        8.0  97.105.19.61  Hampton  2015-09-22  2016-02-06  
4        8.0  97.105.19.61  Hampton  2015-09-22  2016-02-06

In [26]:
result.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 674618 entries, 0 to 674617
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   date         674618 non-null  object 
 1   time         674618 non-null  object 
 2   page_viewed  674618 non-null  object 
 3   user_id      674618 non-null  int64  
 4   cohort_id    674618 non-null  float64
 5   ip           674618 non-null  object 
 6   name         674618 non-null  object 
 7   start_date   674618 non-null  object 
 8   end_date     674618 non-null  object 
dtypes: float64(1), int64(1), object(7)
memory usage: 51.5+ MB


In [27]:
result.cohort_id = result.cohort_id.astype('int')
result['date'] = result.date + " " + result.time
result.drop(columns=('time'), inplace=True)
result.date = pd.to_datetime(result.date)
result = result.set_index('date')

In [28]:
result.head()

page_viewed  user_id  cohort_id  \
date                                                                           
2018-01-26 09:55:03                                    /        1          8   
2018-01-26 09:56:02                              java-ii        1          8   
2018-01-26 09:56:05  java-ii/object-oriented-programming        1          8   
2018-01-26 09:56:06   slides/object_oriented_programming        1          8   
2018-01-26 10:40:15               javascript-i/functions        1          8   

                               ip     name  start_date    end_date  
date                                                                
2018-01-26 09:55:03  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:02  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:05  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:06  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 10:40:15  97.105.19.61  Hampton  2015-09-22  2016-02-06

In [29]:
result.name.value_counts()

Staff         60315
Ceres         40168
Zion          37548
Fortuna       36047
Voyageurs     35624
Ganymede      32447
Apex          31670
Teddy         30328
Deimos        29688
Hyperion      28354
Betelgeuse    28333
Ulysses       28329
Xanadu        27637
Europa        27518
Wrangell      25427
Andromeda     25253
Bayes         25181
Darden        22425
Yosemite      20447
Curie         20410
Bash          11774
Lassen         9495
Arches         8884
Jupiter        8718
Sequoia        7444
Olympic        4925
Kings          2733
Pinnacles      2005
Hampton        1712
Quincy         1165
Niagara         743
Mammoth         691
Glacier         495
Joshua          302
Ike             208
Badlands         93
Franklin         72
Apollo            5
Denali            4
Everglades        1
Name: name, dtype: int64

In [30]:
pd.crosstab(result.name, result.page_viewed)

page_viewed  %20https://github.com/RaulCPena  \
name                                           
Andromeda                                  0   
Apex                                       0   
Apollo                                     0   
Arches                                     0   
Badlands                                   0   
Bash                                       0   
Bayes                                      0   
Betelgeuse                                 0   
Ceres                                      0   
Curie                                      1   
Darden                                     0   
Deimos                                     0   
Denali                                     0   
Europa                                     0   
Everglades                                 0   
Fortuna                                    0   
Franklin                                   0   
Ganymede                                   0   
Glacier                                    0   
Hampton                                    0   
Hyperion                                   0   
Ike                                        0   
Joshua                                     0   
Jupiter                                    0   
Kings                                      0   
Lassen                                     0   
Mammoth                                    0   
Niagara                                    0   
Olympic                                    0   
Pinnacles                                  0   
Quincy                                     0   
Sequoia                                    0   
Staff                                      0   
Teddy                                      0   
Ulysses                                    0   
Voyageurs                                  0   
Wrangell                                   0   
Xanadu                                     0   
Yosemite                                   0   
Zion                                       0   

page_viewed  ,%20https://github.com/RaulCPena  .git  .gitignore  \
name                                                              
Andromeda                                   0     0           0   
Apex                                        0     0           0   
Apollo                                      0     0           0   
Arches                                      0     0           0   
Badlands                                    0     0           0   
Bash                                        0     0           0   
Bayes                                       0     0           0   
Betelgeuse                                  0     0           0   
Ceres                                       0     0           0   
Curie                                       1     0           0   
Darden                                      0     0           0   
Deimos                                      0     0           0   
Denali                                      0     0           0   
Europa                                      0     0           0   
Everglades                                  0     0           0   
Fortuna                                     0     0           0   
Franklin                                    0     0           0   
Ganymede                                    0     0           0   
Glacier                                     0     0           0   
Hampton                                     0     0           0   
Hyperion                                    0     0           0   
Ike                                         0     0           0   
Joshua                                      0     0           0   
Jupiter                                     0     0           0   
Kings                                       0     0           0   
Lassen                                      0     0           0   
Mammoth                                     0     0           0   
Niagara                                 

In [31]:
data_science = result[(result.name == 'Curie') | (result.name == 'Bayes') | (result.name == 'Ada') | (result.name == 'Darden')]

In [32]:
data_science.head()

page_viewed  user_id  cohort_id            ip   name  \
date                                                                       
2019-08-20 09:39:58           /      466         34  97.105.19.58  Bayes   
2019-08-20 09:39:59           /      467         34  97.105.19.58  Bayes   
2019-08-20 09:39:59           /      468         34  97.105.19.58  Bayes   
2019-08-20 09:40:02           /      469         34  97.105.19.58  Bayes   
2019-08-20 09:40:08           /      470         34  97.105.19.58  Bayes   

                     start_date    end_date  
date                                         
2019-08-20 09:39:58  2019-08-19  2020-01-30  
2019-08-20 09:39:59  2019-08-19  2020-01-30  
2019-08-20 09:39:59  2019-08-19  2020-01-30  
2019-08-20 09:40:02  2019-08-19  2020-01-30  
2019-08-20 09:40:08  2019-08-19  2020-01-30

In [33]:
data_science.shape

(68016, 7)

In [34]:
web_dev = result[(result.name != 'Curie') & (result.name != 'Bayes') & (result.name != 'Ada') & (result.name != 'Darden')]

In [35]:
web_dev.shape

(606602, 7)

In [36]:
web_dev.name.value_counts()

Staff         60315
Ceres         40168
Zion          37548
Fortuna       36047
Voyageurs     35624
Ganymede      32447
Apex          31670
Teddy         30328
Deimos        29688
Hyperion      28354
Betelgeuse    28333
Ulysses       28329
Xanadu        27637
Europa        27518
Wrangell      25427
Andromeda     25253
Yosemite      20447
Bash          11774
Lassen         9495
Arches         8884
Jupiter        8718
Sequoia        7444
Olympic        4925
Kings          2733
Pinnacles      2005
Hampton        1712
Quincy         1165
Niagara         743
Mammoth         691
Glacier         495
Joshua          302
Ike             208
Badlands         93
Franklin         72
Apollo            5
Denali            4
Everglades        1
Name: name, dtype: int64

In [37]:
pd.crosstab(data_science.name, data_science.page_viewed)

page_viewed  %20https://github.com/RaulCPena  \
name                                           
Bayes                                      0   
Curie                                      1   
Darden                                     0   

page_viewed  ,%20https://github.com/RaulCPena     /  1-fundamentals  \
name                                                                  
Bayes                                       0  1842               0   
Curie                                       1  1523               0   
Darden                                      0  2041              10   

page_viewed  1-fundamentals/1.1-intro-to-data-science  \
name                                                    
Bayes                                             615   
Curie                                             461   
Darden                                            460   

page_viewed  1-fundamentals/1.2-data-science-pipeline  \
name                                                    
Bayes                                             221   
Curie                                             146   
Darden                                            271   

page_viewed  1-fundamentals/1.3-pipeline-demo  \
name                                            
Bayes                                     130   
Curie                                      60   
Darden                                    110   

page_viewed  1-fundamentals/2.1-excel-overview  \
name                                             
Bayes                                       86   
Curie                                       46   
Darden                                      55   

page_viewed  1-fundamentals/2.1-spreadsheets-overview  \
name                                                    
Bayes                                               4   
Curie                                              73   
Darden                                             16   

page_viewed  1-fundamentals/2.2-excel-functions  ...  \
name                                             ...   
Bayes                                       153  ...   
Curie                                        27  ...   
Darden                                       35  ...   

page_viewed  timeseries/modeling-lesson2  timeseries/overview  \
name                                                            
Bayes                                  3                    1   
Curie                                 11                    8   
Darden                                33                   68   

page_viewed  timeseries/prep  timeseries/project  timeseries/sarimax  \
name                                                                   
Bayes                      6                   0                   1   
Curie                     24                   1                   2   
Darden                    93                  69                   4   

page_viewed  timeseries/svr  timeseries/working-with-time-series-data  \
name                                                                    
Bayes                     1                                         2   
Curie                     1                                         1   
Darden                    6                                        83   

page_viewed  timeseries/working-with-time-series-data-with-pandas  toc  \
name                                                                     
Bayes                                                        2       9   
Curie                                                        2       0   
Darden                                                       4       0   

page_viewed  working-with-time-series-data  
name                                        
Bayes                                    0  
Curie                                    0  
Darden                                  14  

[3 rows x 596 columns]

In [38]:
data_science.name.value_counts()

Bayes     25181
Darden    22425
Curie     20410
Name: name, dtype: int64

In [39]:
result[result.name == 'Ada']

Empty DataFrame
Columns: [page_viewed, user_id, cohort_id, ip, name, start_date, end_date]
Index: []

In [40]:
data_science.groupby(['name','user_id']).page_viewed.value_counts()

name    user_id  page_viewed                                      
Bayes   358      search/search_index.json                             63
                 6-regression/1-overview                              28
                 10-anomaly-detection/1-overview                      22
                 10-anomaly-detection/AnomalyDetectionCartoon.jpeg    22
                 5-stats/3-probability-distributions                  19
                                                                      ..
Darden  785      sql/mysql-overview                                    1
                 timeseries/modeling-lesson1                           1
                 timeseries/prep                                       1
                 timeseries/project                                    1
                 timeseries/working-with-time-series-data              1
Name: page_viewed, Length: 10383, dtype: int64

# 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?


In [41]:
bayes = data_science[data_science.name == 'Bayes']

In [42]:
bayes.page_viewed.value_counts().head(20)

/                                                    1842
1-fundamentals/modern-data-scientist.jpg              626
1-fundamentals/AI-ML-DL-timeline.jpg                  624
1-fundamentals/1.1-intro-to-data-science              615
search/search_index.json                              551
6-regression/1-overview                               521
10-anomaly-detection/AnomalyDetectionCartoon.jpeg     386
10-anomaly-detection/1-overview                       383
6-regression/5.0-evaluate                             333
5-stats/3-probability-distributions                   320
5-stats/4.2-compare-means                             315
appendix/cli-git-overview                             311
6-regression/7.0-model                                310
6-regression/4.0-explore                              267
6-regression/3.0-split-and-scale                      260
7-classification/3-prep                               256
4-python/7.4.3-dataframes                             251
7-classificati

In [43]:
darden = data_science[data_science.name == 'Darden']

In [44]:
darden.page_viewed.value_counts().head(20)

/                                           2041
classification/overview                      759
classification/scale_features_or_not.svg     590
sql/mysql-overview                           513
1-fundamentals/AI-ML-DL-timeline.jpg         470
1-fundamentals/modern-data-scientist.jpg     470
1-fundamentals/1.1-intro-to-data-science     460
stats/compare-means                          338
classification/logistic-regression           334
classification/prep                          321
search/search_index.json                     300
1-fundamentals/DataToAction_v2.jpg           284
classification/explore                       282
classification/evaluation                    280
1-fundamentals/1.2-data-science-pipeline     271
classification/project                       252
classification/acquire                       252
stats/probability-distributions              246
python/data-types-and-variables              235
stats/correlation                            234
Name: page_viewed, d

In [45]:
curie = data_science[data_science.name == 'Curie']

In [46]:
curie.page_viewed.value_counts().head(20)

/                                                    1523
6-regression/1-overview                               595
search/search_index.json                              480
1-fundamentals/modern-data-scientist.jpg              467
1-fundamentals/AI-ML-DL-timeline.jpg                  465
1-fundamentals/1.1-intro-to-data-science              461
3-sql/1-mysql-overview                                441
10-anomaly-detection/1-overview                       345
10-anomaly-detection/AnomalyDetectionCartoon.jpeg     345
4-python/8.4.3-dataframes                             260
4-python/8.4.4-advanced-dataframes                    246
4-python/3-data-types-and-variables                   234
4-python/5-functions                                  203
5-stats/4.2-compare-means                             197
5-stats/2-simulation                                  193
appendix/cli-git-overview                             190
3-sql/7-functions                                     185
7-classificati

In [47]:
bayes.head()

page_viewed  user_id  cohort_id            ip   name  \
date                                                                       
2019-08-20 09:39:58           /      466         34  97.105.19.58  Bayes   
2019-08-20 09:39:59           /      467         34  97.105.19.58  Bayes   
2019-08-20 09:39:59           /      468         34  97.105.19.58  Bayes   
2019-08-20 09:40:02           /      469         34  97.105.19.58  Bayes   
2019-08-20 09:40:08           /      470         34  97.105.19.58  Bayes   

                     start_date    end_date  
date                                         
2019-08-20 09:39:58  2019-08-19  2020-01-30  
2019-08-20 09:39:59  2019-08-19  2020-01-30  
2019-08-20 09:39:59  2019-08-19  2020-01-30  
2019-08-20 09:40:02  2019-08-19  2020-01-30  
2019-08-20 09:40:08  2019-08-19  2020-01-30

In [48]:
web_dev.name.value_counts()

Staff         60315
Ceres         40168
Zion          37548
Fortuna       36047
Voyageurs     35624
Ganymede      32447
Apex          31670
Teddy         30328
Deimos        29688
Hyperion      28354
Betelgeuse    28333
Ulysses       28329
Xanadu        27637
Europa        27518
Wrangell      25427
Andromeda     25253
Yosemite      20447
Bash          11774
Lassen         9495
Arches         8884
Jupiter        8718
Sequoia        7444
Olympic        4925
Kings          2733
Pinnacles      2005
Hampton        1712
Quincy         1165
Niagara         743
Mammoth         691
Glacier         495
Joshua          302
Ike             208
Badlands         93
Franklin         72
Apollo            5
Denali            4
Everglades        1
Name: name, dtype: int64

In [49]:
ceres = web_dev[web_dev.name == 'Ceres']

In [50]:
ceres.page_viewed.value_counts().head(20)

/                                                                            1620
search/search_index.json                                                     1376
javascript-i                                                                  977
toc                                                                           909
html-css                                                                      753
java-iii                                                                      674
java-ii                                                                       667
jquery                                                                        632
mysql                                                                         617
spring                                                                        546
javascript-ii                                                                 519
java-i                                                                        510
html-css/css-i/f

In [51]:
zion = web_dev[web_dev.name == 'Zion']

In [52]:
zion.page_viewed.value_counts().head(20)

/                                                                            1756
toc                                                                          1457
javascript-i                                                                  868
java-iii                                                                      742
search/search_index.json                                                      689
spring                                                                        650
html-css                                                                      649
javascript-ii                                                                 637
java-ii                                                                       613
mysql                                                                         598
java-i                                                                        593
jquery                                                                        559
spring/fundament

In [53]:
fortuna = web_dev[web_dev.name == 'Fortuna']

In [54]:
fortuna.page_viewed.value_counts().head(20)

/                                    1962
toc                                  1273
search/search_index.json              989
java-iii                              767
javascript-i                          756
java-ii                               637
spring                                616
html-css                              578
mysql                                 571
java-i                                538
jquery                                501
javascript-ii                         482
java-iii/servlets                     416
java-iii/jsp-and-jstl                 402
mysql/tables                          373
java-i/syntax-types-and-variables     358
java-i/introduction-to-java           356
mysql/basic-statements                348
spring/fundamentals/controllers       343
appendix                              337
Name: page_viewed, dtype: int64

In [55]:
voyageurs = web_dev[web_dev.name == 'Voyageurs']

In [56]:
voyageurs.page_viewed.value_counts().head(20)

/                                      2098
javascript-i                            884
java-iii                                770
java-ii                                 754
mysql                                   663
spring                                  650
java-i                                  639
javascript-ii                           584
jquery                                  583
html-css                                528
java-i/introduction-to-java             447
mysql/databases                         439
mysql/tables                            437
mysql/users                             410
java-iii/servlets                       397
appendix                                393
mysql/basic-statements                  384
java-ii/object-oriented-programming     380
java-iii/jsp-and-jstl                   367
javascript-i/javascript-with-html       354
Name: page_viewed, dtype: int64

In [57]:
ganymede = web_dev[web_dev.name == 'Ganymede']

In [58]:
ganymede.page_viewed.value_counts().head(20)

/                                      1618
search/search_index.json               1050
toc                                     810
javascript-i                            694
java-iii                                620
java-ii                                 607
appendix                                526
jquery                                  496
javascript-ii                           483
java-i                                  482
mysql                                   474
html-css                                472
spring                                  465
java-iii/servlets                       328
java-i/syntax-types-and-variables       317
java-iii/jsp-and-jstl                   308
java-ii/collections                     302
java-ii/object-oriented-programming     298
java-ii/arrays                          286
mysql/databases                         282
Name: page_viewed, dtype: int64

**for data science it looks like fundamentals is the most common. For web dev it looks like java and javascript are the most common.**

# - 2. Is there a cohort that referred to a lesson significantly more that other cohorts seemed to gloss over? 

**Curie appeared to access the Python modules significantly more than other data science cohorts**

In [59]:
apex = web_dev[web_dev.name == 'Apex']

In [60]:
apex.page_viewed.value_counts().head(20)

search/search_index.json               1361
/                                      1244
toc                                     799
html-css                                708
java-iii                                616
javascript-i                            610
java-ii                                 595
spring                                  572
java-i                                  528
mysql                                   445
jquery                                  437
javascript-ii                           431
appendix                                426
java-i/syntax-types-and-variables       346
java-iii/servlets                       338
java-ii/object-oriented-programming     334
mysql/tables                            332
java-ii/arrays                          308
mysql/databases                         306
java-ii/collections                     294
Name: page_viewed, dtype: int64

In [61]:
teddy = web_dev[web_dev.name == 'Teddy']

In [62]:
teddy.page_viewed.value_counts().head(20)

/                                   1754
java-iii                             712
spring                               707
mysql                                631
mkdocs/search_index.json             595
javascript-i                         589
java-i                               501
jquery                               498
java-ii                              494
appendix                             488
javascript-ii                        468
mysql/tables                         387
mysql/databases                      382
javascript-i/functions               360
mysql/basic-statements               360
mysql/users                          356
javascript-i/loops                   337
javascript-i/conditionals            337
javascript-ii/promises               311
spring/fundamentals/repositories     287
Name: page_viewed, dtype: int64

In [63]:
deimos = web_dev[web_dev.name == 'Deimos']

In [64]:
deimos.page_viewed.value_counts().head(20)

/                                              1319
javascript-i                                    696
search/search_index.json                        662
html-css                                        609
toc                                             569
java-iii                                        517
spring                                          512
java-ii                                         506
mysql                                           498
jquery                                          466
java-i                                          434
javascript-ii                                   392
mysql/tables                                    371
mysql/databases                                 359
java-iii/jsp-and-jstl                           349
mysql/users                                     342
java-iii/servlets                               326
html-css/css-ii/bootstrap-introduction          325
html-css/css-i/flexbox/flexbox-fundamentals     324
mysql/basic-

In [65]:
hyperion = web_dev[web_dev.name == 'Hyperion']

In [66]:
hyperion.page_viewed.value_counts().head(20)

/                                                                            1245
toc                                                                           977
javascript-i                                                                  884
java-iii                                                                      651
java-ii                                                                       632
search/search_index.json                                                      630
mysql                                                                         517
jquery                                                                        490
java-i                                                                        444
spring                                                                        442
html-css                                                                      418
javascript-ii                                                                 410
html-css/css-i/s

Betelgeuse    28333
Ulysses       28329
Xanadu        27637
Europa        27518
Wrangell      25427
Andromeda     25253
Yosemite      20447
Bash          11774
Lassen         9495
Arches         8884
Jupiter        8718
Sequoia        7444
Olympic        4925
Kings          2733
Pinnacles      2005
Hampton        1712
Quincy         1165
Niagara         743
Mammoth         691
Glacier         495
Joshua          302
Ike             208
Badlands         93
Franklin         72
Apollo            5
Denali            4
Everglades        1

In [67]:
betelgeuse = web_dev[web_dev.name == 'Betelgeuse']

In [68]:
betelgeuse.page_viewed.value_counts().head(20)

/                                                                            868
search/search_index.json                                                     718
javascript-i                                                                 686
toc                                                                          577
jquery                                                                       540
html-css/elements                                                            472
java-ii                                                                      444
java-iii                                                                     444
html-css                                                                     425
java-i                                                                       422
html-css/css-ii/bootstrap-grid-system                                        414
javascript-ii                                                                374
javascript-i/javascript-with

In [69]:
ulysses = web_dev[web_dev.name == 'Ulysses']

In [70]:
ulysses.page_viewed.value_counts().head(20)

/                                                                            1618
mkdocs/search_index.json                                                      721
html-css                                                                      555
javascript-i                                                                  523
java-ii                                                                       423
java-iii                                                                      411
spring                                                                        396
java-i                                                                        375
jquery                                                                        363
mysql                                                                         361
spring/fundamentals/form-model-binding                                        357
html-css/css-ii/bootstrap-introduction                                        351
mysql/users     

In [71]:
xanadu = web_dev[web_dev.name == 'Xanadu']

In [72]:
xanadu.page_viewed.value_counts().head(20)

/                                    916
javascript-i                         718
html-css                             587
search/search_index.json             576
jquery                               543
mysql                                507
java-ii                              501
java-iii                             495
javascript-ii                        493
java-i                               483
spring                               453
toc                                  435
html-css/elements                    433
mysql/tables                         373
appendix                             320
mysql/databases                      318
javascript-i/functions               305
mysql/basic-statements               301
java-iii/jsp-and-jstl                297
java-i/syntax-types-and-variables    291
Name: page_viewed, dtype: int64

In [73]:
europa = web_dev[web_dev.name == 'Europa']

In [74]:
europa.page_viewed.value_counts().head(20)

/                                                                            1223
toc                                                                           949
search/search_index.json                                                      734
javascript-i                                                                  478
java-iii                                                                      443
html-css/elements                                                             422
mysql                                                                         367
java-ii                                                                       366
jquery                                                                        360
html-css/css-i/selectors-and-properties                                       351
html-css                                                                      329
mysql/tables                                                                  328
mysql/databases 

In [75]:
result.head()

page_viewed  user_id  cohort_id  \
date                                                                           
2018-01-26 09:55:03                                    /        1          8   
2018-01-26 09:56:02                              java-ii        1          8   
2018-01-26 09:56:05  java-ii/object-oriented-programming        1          8   
2018-01-26 09:56:06   slides/object_oriented_programming        1          8   
2018-01-26 10:40:15               javascript-i/functions        1          8   

                               ip     name  start_date    end_date  
date                                                                
2018-01-26 09:55:03  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:02  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:05  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:06  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 10:40:15  97.105.19.61  Hampton  2015-09-22  2016-02-06

In [76]:
pd.crosstab(web_dev.page_viewed, web_dev.name)

name                           Andromeda  Apex  Apollo  Arches  Badlands  \
page_viewed                                                                
.git                                   0     0       0       0         0   
.gitignore                             0     0       0       0         0   
.well-known/assetlinks.json            0     0       0       0         0   
/                                   1156  1244       1     622        17   
00_                                    0     0       0       0         0   
...                                  ...   ...     ...     ...       ...   
web-design/ux/purpose                  8    22       0       0         0   
web-dev-day-two                        0     0       0       0         0   
working-with-time-series-data          0     0       0       0         0   
wp-admin                               0     0       0       0         0   
wp-login                               0     0       0       0         0   

name                           Bash  Betelgeuse  Ceres  Deimos  Denali  ...  \
page_viewed                                                             ...   
.git                              0           0      0       0       0  ...   
.gitignore                        0           0      0       0       0  ...   
.well-known/assetlinks.json       0           0      0       0       0  ...   
/                               532         868   1620    1319       1  ...   
00_                               0           0      0       0       0  ...   
...                             ...         ...    ...     ...     ...  ...   
web-design/ux/purpose             0          32     28      11       0  ...   
web-dev-day-two                   2           0      0       0       0  ...   
working-with-time-series-data     0           0      0       0       0  ...   
wp-admin                          0           0      0       0       0  ...   
wp-login                          0           0      0       0       0  ...   

name                           Quincy  Sequoia  Staff  Teddy  Ulysses  \
page_viewed                                                             
.git                                0        0      0      0        0   
.gitignore                          0        0      0      0        0   
.well-known/assetlinks.json         0        0      0      0        0   
/                                 138      630   4633   1754     1618   
00_                                 0        0      1      0        0   
...                               ...      ...    ...    ...      ...   
web-design/ux/purpose               0        1     13      0        0   
web-dev-day-two                     0        0      0      0        0   
working-with-time-series-data       0        0      1      0        0   
wp-admin                            0        0      0      1        0   
wp-login                            0        0      0      1        0   

name                           Voyageurs  Wrangell  Xanadu  Yosemite  Zion  
page_viewed                                                                 
.git                                   1         0       0         0     0  
.gitignore                             1         0       0         0     0  
.well-known/assetlinks.json            0         0       0         0     0  
/                                   2098      1112     916       962  1756  
00_                                    0         0       0         0     0  
...                                  ...       ...     ...       ...   ...  
web-design/ux/purpose                  1         1       2         2     1  
web-dev-day-two                        0         0       0         0     0  
working-with-time-series-data          0         0       0         0     0  
wp-admin                               0         0       0         0     0  
wp-login                               0         0       0         0     0  

[1939 rows x 37 columns]

In [77]:
wrangell =  web_dev[web_dev.name == 'Wrangell']

In [78]:
wrangell.page_viewed.value_counts().head(20)

/                                                                            1112
toc                                                                           990
javascript-i                                                                  553
search/search_index.json                                                      489
html-css                                                                      405
java-i                                                                        401
jquery                                                                        374
java-iii                                                                      335
java-ii                                                                       323
javascript-i/loops                                                            312
html-css/css-ii/bootstrap-grid-system                                         312
javascript-i/functions                                                        312
html-css/css-ii/

In [79]:
andromeda = web_dev[web_dev.name == 'Andromeda']

In [80]:
andromeda.page_viewed.value_counts().head(20)

/                                                                            1156
toc                                                                           637
javascript-i                                                                  509
spring                                                                        485
java-iii                                                                      430
java-ii                                                                       395
html-css                                                                      372
mysql                                                                         346
java-i                                                                        336
jquery                                                                        335
search/search_index.json                                                      318
mysql/tables                                                                  316
spring/fundament

In [81]:
yosemite = web_dev[web_dev.name == 'Yosemite']

In [82]:
yosemite.page_viewed.value_counts().head(20)

/                                        962
toc                                      686
javascript-i                             437
spring                                   374
html-css                                 362
search/search_index.json                 361
java-iii                                 340
javascript-ii                            332
mysql                                    318
javascript-ii/map-filter-reduce          298
jquery                                   278
mysql/tables                             277
java-ii                                  268
javascript-ii/promises                   249
javascript-i/loops                       248
java-i                                   245
html-css/css-ii/bootstrap-grid-system    242
mysql/users                              232
javascript-i/functions                   230
javascript-i/javascript-with-html        227
Name: page_viewed, dtype: int64

**Not seeing any variance in web dev**

# Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?

In [83]:
data_science.head()

page_viewed  user_id  cohort_id            ip   name  \
date                                                                       
2019-08-20 09:39:58           /      466         34  97.105.19.58  Bayes   
2019-08-20 09:39:59           /      467         34  97.105.19.58  Bayes   
2019-08-20 09:39:59           /      468         34  97.105.19.58  Bayes   
2019-08-20 09:40:02           /      469         34  97.105.19.58  Bayes   
2019-08-20 09:40:08           /      470         34  97.105.19.58  Bayes   

                     start_date    end_date  
date                                         
2019-08-20 09:39:58  2019-08-19  2020-01-30  
2019-08-20 09:39:59  2019-08-19  2020-01-30  
2019-08-20 09:39:59  2019-08-19  2020-01-30  
2019-08-20 09:40:02  2019-08-19  2020-01-30  
2019-08-20 09:40:08  2019-08-19  2020-01-30

In [84]:
darden.user_id.value_counts() < 20

685    False
698    False
689    False
699    False
681    False
692    False
688    False
691    False
682    False
678    False
696    False
684    False
680    False
268    False
687    False
686    False
695    False
690    False
694    False
739    False
693    False
683    False
781    False
783    False
780    False
785    False
697     True
679     True
Name: user_id, dtype: bool

In [85]:
user1 = darden[darden.user_id == 697]

In [86]:
user1.shape

(13, 7)

In [87]:
user1

page_viewed  user_id  \
date                                                                     
2020-07-13 15:20:27                                         /      697   
2020-07-13 15:20:48                    3-sql/1-mysql-overview      697   
2020-07-13 15:20:50  1-fundamentals/1.1-intro-to-data-science      697   
2020-07-13 15:20:50  1-fundamentals/modern-data-scientist.jpg      697   
2020-07-13 15:20:50      1-fundamentals/AI-ML-DL-timeline.jpg      697   
2020-07-13 15:20:59  1-fundamentals/1.2-data-science-pipeline      697   
2020-07-13 15:20:59        1-fundamentals/DataToAction_v2.jpg      697   
2020-07-13 15:21:01  1-fundamentals/1.1-intro-to-data-science      697   
2020-07-13 15:21:02      1-fundamentals/AI-ML-DL-timeline.jpg      697   
2020-07-13 15:21:02  1-fundamentals/modern-data-scientist.jpg      697   
2020-07-13 15:21:12                  2-storytelling/3-tableau      697   
2020-07-13 15:21:16        2-storytelling/3.3-creating-charts      697   
2020-07-13 15:26:39            4-python/8.4.1-pandas-overview      697   

                     cohort_id            ip    name  start_date    end_date  
date                                                                          
2020-07-13 15:20:27         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:20:48         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:20:50         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:20:50         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:20:50         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:20:59         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:20:59         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:21:01         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:21:02         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:21:02         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:21:12         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:21:16         59  136.50.70.27  Darden  2020-07-13  2021-01-12  
2020-07-13 15:26:39         59  136.50.70.27  Darden  2020-07-13  2021-01-12

In [88]:
user2 = darden[darden.user_id == 679]

In [89]:
user2.shape

(11, 7)

In [90]:
user2

page_viewed  user_id  \
date                                                                     
2020-07-13 14:37:22                                         /      679   
2020-07-13 14:39:21            13-advanced-topics/1-tidy-data      679   
2020-07-13 14:39:36  1-fundamentals/1.1-intro-to-data-science      679   
2020-07-13 14:39:37      1-fundamentals/AI-ML-DL-timeline.jpg      679   
2020-07-13 14:39:37  1-fundamentals/modern-data-scientist.jpg      679   
2020-07-13 15:49:31  1-fundamentals/1.1-intro-to-data-science      679   
2020-07-13 15:49:32  1-fundamentals/modern-data-scientist.jpg      679   
2020-07-13 15:49:32      1-fundamentals/AI-ML-DL-timeline.jpg      679   
2020-07-14 08:05:15  1-fundamentals/1.1-intro-to-data-science      679   
2020-07-14 08:05:15      1-fundamentals/AI-ML-DL-timeline.jpg      679   
2020-07-14 08:05:15  1-fundamentals/modern-data-scientist.jpg      679   

                     cohort_id             ip    name  start_date    end_date  
date                                                                           
2020-07-13 14:37:22         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:21         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:36         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:37         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:37         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 15:49:31         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 15:49:32         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 15:49:32         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-14 08:05:15         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-14 08:05:15         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-14 08:05:15         59  24.28.146.155  Darden  2020-07-13  2021-01-12

In [91]:
curie.user_id.value_counts() < 20

581    False
576    False
590    False
584    False
580    False
582    False
579    False
585    False
586    False
589    False
617    False
591    False
578    False
588    False
616    False
575    False
587    False
583    False
577    False
746     True
787     True
Name: user_id, dtype: bool

In [92]:
user3 = curie[curie.user_id == 746]

In [93]:
user3.shape

(1, 7)

In [94]:
user3

page_viewed  user_id  cohort_id               ip   name  \
date                                                                          
2020-09-10 10:50:28           /      746         55  173.175.100.201  Curie   

                     start_date    end_date  
date                                         
2020-09-10 10:50:28  2020-02-03  2020-07-07

In [95]:
user4 = curie[curie.user_id == 787]

In [96]:
user4

page_viewed  user_id  \
date                                                                  
2020-10-29 22:08:03  appendix/interview_questions_students      787   

                     cohort_id              ip   name  start_date    end_date  
date                                                                           
2020-10-29 22:08:03         55  99.126.113.140  Curie  2020-02-03  2020-07-07

In [97]:
bayes.user_id.value_counts() < 20

485    False
475    False
476    False
479    False
478    False
482    False
471    False
469    False
466    False
473    False
481    False
358    False
484    False
480    False
483    False
472    False
468    False
474    False
467    False
470    False
477    False
487     True
650     True
Name: user_id, dtype: bool

In [98]:
user5 = bayes[bayes.user_id == 487]

In [99]:
user5

page_viewed  user_id  \
date                                                                          
2019-08-30 12:01:29                                              /      487   
2019-08-30 12:02:11                           appendix/cli/1-intro      487   
2019-08-30 12:11:10                   appendix/cli/2-listing-files      487   
2019-08-30 13:37:40                        appendix/cli/0-overview      487   
2019-08-30 13:37:58                      appendix/cli-git-overview      487   
2019-08-30 14:12:22                      appendix/cli/3-file-paths      487   
2019-08-30 14:19:45       appendix/cli/4-navigating-the-filesystem      487   
2019-08-30 14:32:22  appendix/cli/5-creating-files-and-directories      487   
2019-08-30 14:38:57       appendix/cli/4-navigating-the-filesystem      487   
2019-08-30 14:59:46  appendix/cli/5-creating-files-and-directories      487   
2019-08-30 15:15:06                    appendix/cli/6-moving-files      487   
2019-08-30 15:17:17  appendix/cli/5-creating-files-and-directories      487   
2019-08-30 15:20:51  appendix/cli/5-creating-files-and-directories      487   
2019-08-30 15:21:02  appendix/cli/5-creating-files-and-directories      487   
2019-08-30 15:21:16       appendix/cli/4-navigating-the-filesystem      487   
2019-08-30 15:21:58                    appendix/cli/6-moving-files      487   
2019-08-30 15:43:38                    appendix/cli/6-moving-files      487   

                     cohort_id            ip   name  start_date    end_date  
date                                                                         
2019-08-30 12:01:29         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 12:02:11         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 12:11:10         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 13:37:40         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 13:37:58         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 14:12:22         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 14:19:45         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 14:32:22         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 14:38:57         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 14:59:46         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 15:15:06         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 15:17:17         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 15:20:51         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 15:21:02         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 15:21:16         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 15:21:58         34  97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-30 15:43:38         34  97.105.19.58  Bayes  2019-08-19  2020-01-30

In [100]:
user6 = bayes[bayes.user_id == 650]

In [101]:
user6

page_viewed  \
date                                                                     
2020-05-07 10:35:25                                                  /   
2020-05-07 10:35:52  appendix/professional-development/virtual-meet...   
2020-06-25 10:18:26                                                  /   
2020-06-25 10:19:01                          appendix/cli-git-overview   
2020-06-25 10:20:00                      appendix/ds-environment-setup   
2020-07-03 15:13:38                                                  /   
2020-07-03 15:15:21                                   11-nlp/3-acquire   
2020-07-06 11:41:32                      appendix/ds-environment-setup   
2020-08-03 11:18:31                                                  /   
2020-08-09 13:17:01                                                  /   
2020-08-09 13:18:23                         python/intro-to-matplotlib   
2020-08-09 14:06:52                         python/intro-to-matplotlib   
2020-09-30 07:04:03                                                  /   

                     user_id  cohort_id             ip   name  start_date  \
date                                                                        
2020-05-07 10:35:25      650         34  70.123.225.30  Bayes  2019-08-19   
2020-05-07 10:35:52      650         34  70.123.225.30  Bayes  2019-08-19   
2020-06-25 10:18:26      650         34  70.123.225.30  Bayes  2019-08-19   
2020-06-25 10:19:01      650         34  70.123.225.30  Bayes  2019-08-19   
2020-06-25 10:20:00      650         34  70.123.225.30  Bayes  2019-08-19   
2020-07-03 15:13:38      650         34  70.123.225.30  Bayes  2019-08-19   
2020-07-03 15:15:21      650         34  70.123.225.30  Bayes  2019-08-19   
2020-07-06 11:41:32      650         34  70.123.225.30  Bayes  2019-08-19   
2020-08-03 11:18:31      650         34  70.123.225.30  Bayes  2019-08-19   
2020-08-09 13:17:01      650         34  70.123.225.30  Bayes  2019-08-19   
2020-08-09 13:18:23      650         34  70.123.225.30  Bayes  2019-08-19   
2020-08-09 14:06:52      650         34  70.123.225.30  Bayes  2019-08-19   
2020-09-30 07:04:03      650         34  70.123.225.30  Bayes  2019-08-19   

                       end_date  
date                             
2020-05-07 10:35:25  2020-01-30  
2020-05-07 10:35:52  2020-01-30  
2020-06-25 10:18:26  2020-01-30  
2020-06-25 10:19:01  2020-01-30  
2020-06-25 10:20:00  2020-01-30  
2020-07-03 15:13:38  2020-01-30  
2020-07-03 15:15:21  2020-01-30  
2020-07-06 11:41:32  2020-01-30  
2020-08-03 11:18:31  2020-01-30  
2020-08-09 13:17:01  2020-01-30  
2020-08-09 13:18:23  2020-01-30  
2020-08-09 14:06:52  2020-01-30  
2020-09-30 07:04:03  2020-01-30

In [102]:
(web_dev.user_id.value_counts() < 5).tail(30)

401    False
786    False
115    False
461    False
71     False
169    False
399    False
82     False
152    False
246     True
216     True
85      True
177     True
348     True
81      True
573     True
574     True
610     True
97      True
165     True
619     True
163     True
593     True
212     True
652     True
592     True
649     True
66      True
745     True
651     True
Name: user_id, dtype: bool

In [103]:
user7 = web_dev[web_dev.user_id == 651]

In [104]:
user7

page_viewed  user_id  cohort_id            ip      name  \
date                                                                          
2020-05-17 19:17:33           /      651         27  66.25.168.77  Yosemite   

                     start_date    end_date  
date                                         
2020-05-17 19:17:33  2018-11-05  2019-04-03

In [105]:
user8 = web_dev[web_dev.user_id == 745]

In [106]:
user8

page_viewed  user_id  cohort_id  \
date                                                                           
2020-09-02 16:23:23  javascript-i/bom-and-dom/calculator      745         28   

                                ip   name  start_date    end_date  
date                                                               
2020-09-02 16:23:23  72.181.145.92  Staff  2014-02-04  2014-02-04

In [107]:
user9 = web_dev[web_dev.user_id == 66]

In [108]:
user9

page_viewed  user_id  cohort_id              ip   name  \
date                                                                         
2018-01-29 15:48:48           /       66         13  67.198.116.192  Kings   

                     start_date    end_date  
date                                         
2018-01-29 15:48:48  2016-05-23  2016-09-15

In [109]:
user10 = web_dev[web_dev.user_id == 649]

In [110]:
user10

page_viewed  user_id  cohort_id             ip  \
date                                                                 
2020-05-06 19:06:32           /      649          5  70.125.150.41   

                           name  start_date    end_date  
date                                                     
2020-05-06 19:06:32  Everglades  2014-11-18  2015-02-24

In [111]:
user11 = web_dev[web_dev.user_id == 592]

In [112]:
user11

page_viewed  user_id  cohort_id            ip   name  \
date                                                                       
2020-02-07 12:17:45           /      592         28  97.105.19.58  Staff   

                     start_date    end_date  
date                                         
2020-02-07 12:17:45  2014-02-04  2014-02-04

In [113]:
user12 = web_dev[web_dev.user_id == 652]

In [114]:
user12

page_viewed  user_id  cohort_id             ip   name  \
date                                                                        
2020-05-18 09:56:10     loguout      652         28  71.150.217.33  Staff   

                     start_date    end_date  
date                                         
2020-05-18 09:56:10  2014-02-04  2014-02-04

In [115]:
user13 = web_dev[web_dev.user_id == 212]

In [116]:
user13

page_viewed  user_id  cohort_id  \
date                                                                       
2018-06-19 09:21:08  students/units/75/sub_units/268      212          1   

                                  ip    name  start_date    end_date  
date                                                                  
2018-06-19 09:21:08  170.248.173.247  Arches  2014-02-04  2014-04-22

In [117]:
user14 = web_dev[web_dev.user_id == 212]

In [118]:
user14

page_viewed  user_id  cohort_id  \
date                                                                       
2018-06-19 09:21:08  students/units/75/sub_units/268      212          1   

                                  ip    name  start_date    end_date  
date                                                                  
2018-06-19 09:21:08  170.248.173.247  Arches  2014-02-04  2014-04-22

In [119]:
user15 = web_dev[web_dev.user_id == 593]

In [120]:
user15

page_viewed  user_id  cohort_id              ip    name  \
date                                                                          
2020-02-17 11:45:21  index.html      593         14  167.24.104.150  Lassen   

                     start_date    end_date  
date                                         
2020-02-17 11:45:21  2016-07-18  2016-11-10

In [121]:
user16 = web_dev[web_dev.user_id == 163]

In [122]:
user16

page_viewed  user_id  cohort_id               ip    name  \
date                                                                           
2018-03-29 22:40:57           /      163         19  192.171.117.234  Quincy   

                     start_date    end_date  
date                                         
2018-03-29 22:40:57  2017-06-05  2017-09-22

Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses? Any odd user-agents?

In [123]:
darden

page_viewed  user_id  \
date                                                                     
2020-07-13 14:34:44                                         /      678   
2020-07-13 14:37:22                                         /      679   
2020-07-13 14:38:28                                         /      680   
2020-07-13 14:39:21            13-advanced-topics/1-tidy-data      679   
2020-07-13 14:39:36  1-fundamentals/1.1-intro-to-data-science      679   
...                                                       ...      ...   
2020-11-02 15:47:59                 appendix/useful_libraries      698   
2020-11-02 15:48:03             appendix/ds-environment-setup      698   
2020-11-02 15:48:14     appendix/interview_questions_students      698   
2020-11-02 15:50:07                        appendix/open_data      698   
2020-11-02 15:52:23                        timeseries/sarimax      698   

                     cohort_id             ip    name  start_date    end_date  
date                                                                           
2020-07-13 14:34:44         59  76.201.20.193  Darden  2020-07-13  2021-01-12  
2020-07-13 14:37:22         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:38:28         59  136.50.56.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:21         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:36         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
...                        ...            ...     ...         ...         ...  
2020-11-02 15:47:59         59  24.28.150.196  Darden  2020-07-13  2021-01-12  
2020-11-02 15:48:03         59  24.28.150.196  Darden  2020-07-13  2021-01-12  
2020-11-02 15:48:14         59  24.28.150.196  Darden  2020-07-13  2021-01-12  
2020-11-02 15:50:07         59  24.28.150.196  Darden  2020-07-13  2021-01-12  
2020-11-02 15:52:23         59  24.28.150.196  Darden  2020-07-13  2021-01-12  

[22425 rows x 7 columns]

In [124]:
darden.ip.nunique()

155

In [125]:
curie.ip.nunique()

240

In [126]:
curie.index.min()

Timestamp('2020-02-03 15:39:34')

In [127]:
curie.index.max()

Timestamp('2020-11-02 15:57:29')

In [128]:
curie.ip.max()

'99.62.54.30'

In [129]:
darden[darden.index.hour == 3]

page_viewed  user_id  \
date                                                                     
2020-07-16 03:07:52                                         /      689   
2020-08-13 03:03:36                                         /      698   
2020-08-24 03:02:14                  search/search_index.json      689   
2020-08-24 03:15:17    1-fundamentals/2.4-more-excel-features      689   
2020-08-24 03:15:45         1-fundamentals/2.1-excel-overview      689   
2020-08-24 03:16:04        1-fundamentals/2.2-excel-functions      689   
2020-08-24 03:44:29                     storytelling/overview      688   
2020-08-24 03:47:54     fundamentals/visualization-with-excel      688   
2020-08-24 03:49:08                     storytelling/overview      688   
2020-08-24 03:49:30                   storytelling/understand      688   
2020-09-10 03:42:21                                         /      691   
2020-09-10 03:42:32                       classification/prep      691   
2020-09-21 03:26:44                                         /      698   
2020-09-21 03:26:52                            stats/sampling      698   
2020-10-30 03:58:29  timeseries/working-with-time-series-data      688   

                     cohort_id              ip    name  start_date    end_date  
date                                                                            
2020-07-16 03:07:52         59  72.181.127.233  Darden  2020-07-13  2021-01-12  
2020-08-13 03:03:36         59    67.11.186.90  Darden  2020-07-13  2021-01-12  
2020-08-24 03:02:14         59  72.181.127.233  Darden  2020-07-13  2021-01-12  
2020-08-24 03:15:17         59  72.181.127.233  Darden  2020-07-13  2021-01-12  
2020-08-24 03:15:45         59  72.181.127.233  Darden  2020-07-13  2021-01-12  
2020-08-24 03:16:04         59  72.181.127.233  Darden  2020-07-13  2021-01-12  
2020-08-24 03:44:29         59  70.125.137.220  Darden  2020-07-13  2021-01-12  
2020-08-24 03:47:54         59  70.125.137.220  Darden  2020-07-13  2021-01-12  
2020-08-24 03:49:08         59  70.125.137.220  Darden  2020-07-13  2021-01-12  
2020-08-24 03:49:30         59  70.125.137.220  Darden  2020-07-13  2021-01-12  
2020-09-10 03:42:21         59    66.69.83.140  Darden  2020-07-13  2021-01-12  
2020-09-10 03:42:32         59    66.69.83.140  Darden  2020-07-13  2021-01-12  
2020-09-21 03:26:44         59   67.11.185.135  Darden  2020-07-13  2021-01-12  
2020-09-21 03:26:52         59   67.11.185.135  Darden  2020-07-13  2021-01-12  
2020-10-30 03:58:29         59  70.125.137.220  Darden  2020-07-13  2021-01-12

In [130]:
darden.index.hour.value_counts()

9     3279
11    2338
10    2322
15    1790
14    1778
16    1627
13    1445
8     1378
12    1131
17     744
21     712
20     700
19     657
18     648
7      546
22     429
23     248
6      237
0      149
1       96
5       77
2       53
4       26
3       15
Name: date, dtype: int64

In [131]:
curie[curie.index.hour == 3]

page_viewed  user_id  cohort_id  \
date                                                                       
2020-04-26 03:32:38    7-classification/5-evaluation      616         55   
2020-06-04 03:42:50                   11-nlp/project      584         55   
2020-06-04 03:42:50  11-nlp/github_repo_language.gif      584         55   
2020-06-04 03:47:30                 11-nlp/5-explore      584         55   
2020-08-04 03:16:43                                /      590         55   
2020-08-04 03:16:49               sql/mysql-overview      590         55   
2020-08-04 03:19:42                    sql/databases      590         55   
2020-08-04 03:19:56                       sql/tables      590         55   
2020-08-04 03:20:03             sql/clauses-overview      590         55   
2020-08-04 03:20:08                        sql/where      590         55   
2020-08-04 03:20:26                    sql/functions      590         55   

                                 ip   name  start_date    end_date  
date                                                                
2020-04-26 03:32:38    70.114.9.241  Curie  2020-02-03  2020-07-07  
2020-06-04 03:42:50  70.123.236.191  Curie  2020-02-03  2020-07-07  
2020-06-04 03:42:50  70.123.236.191  Curie  2020-02-03  2020-07-07  
2020-06-04 03:47:30  70.123.236.191  Curie  2020-02-03  2020-07-07  
2020-08-04 03:16:43    130.45.47.56  Curie  2020-02-03  2020-07-07  
2020-08-04 03:16:49    130.45.47.56  Curie  2020-02-03  2020-07-07  
2020-08-04 03:19:42    130.45.47.56  Curie  2020-02-03  2020-07-07  
2020-08-04 03:19:56    130.45.47.56  Curie  2020-02-03  2020-07-07  
2020-08-04 03:20:03    130.45.47.56  Curie  2020-02-03  2020-07-07  
2020-08-04 03:20:08    130.45.47.56  Curie  2020-02-03  2020-07-07  
2020-08-04 03:20:26    130.45.47.56  Curie  2020-02-03  2020-07-07

In [132]:
data_science.user_id.value_counts()

581    3251
685    2542
485    2298
475    2135
476    2046
       ... 
697      13
650      13
679      11
746       1
787       1
Name: user_id, Length: 72, dtype: int64

In [133]:
data_science[data_science.user_id == 581]

page_viewed  user_id  \
date                                                                     
2020-02-03 15:40:35                                         /      581   
2020-02-03 15:42:00                                         /      581   
2020-02-03 15:43:47  1-fundamentals/1.1-intro-to-data-science      581   
2020-02-03 15:43:48  1-fundamentals/modern-data-scientist.jpg      581   
2020-02-03 15:43:48      1-fundamentals/AI-ML-DL-timeline.jpg      581   
...                                                       ...      ...   
2020-11-02 15:22:32  timeseries/working-with-time-series-data      581   
2020-11-02 15:22:38                        timeseries/acquire      581   
2020-11-02 15:22:44                           timeseries/prep      581   
2020-11-02 15:33:20                  search/search_index.json      581   
2020-11-02 15:33:27          advanced-topics/cross-validation      581   

                     cohort_id              ip   name  start_date    end_date  
date                                                                           
2020-02-03 15:40:35         55    97.105.19.58  Curie  2020-02-03  2020-07-07  
2020-02-03 15:42:00         55    97.105.19.58  Curie  2020-02-03  2020-07-07  
2020-02-03 15:43:47         55    97.105.19.58  Curie  2020-02-03  2020-07-07  
2020-02-03 15:43:48         55    97.105.19.58  Curie  2020-02-03  2020-07-07  
2020-02-03 15:43:48         55    97.105.19.58  Curie  2020-02-03  2020-07-07  
...                        ...             ...    ...         ...         ...  
2020-11-02 15:22:32         55  70.112.179.142  Curie  2020-02-03  2020-07-07  
2020-11-02 15:22:38         55  70.112.179.142  Curie  2020-02-03  2020-07-07  
2020-11-02 15:22:44         55  70.112.179.142  Curie  2020-02-03  2020-07-07  
2020-11-02 15:33:20         55  70.112.179.142  Curie  2020-02-03  2020-07-07  
2020-11-02 15:33:27         55  70.112.179.142  Curie  2020-02-03  2020-07-07  

[3251 rows x 7 columns]

In [134]:
data_science[data_science.user_id == 685]

page_viewed  user_id  \
date                                                                     
2020-07-13 14:40:31                                         /      685   
2020-07-13 23:23:10                                         /      685   
2020-07-14 08:30:16                                         /      685   
2020-07-14 10:37:01                                         /      685   
2020-07-14 11:09:36  1-fundamentals/1.1-intro-to-data-science      685   
...                                                       ...      ...   
2020-11-02 15:20:36               timeseries/modeling-lesson1      685   
2020-11-02 15:20:50               timeseries/modeling-lesson1      685   
2020-11-02 15:42:31                   classification/overview      685   
2020-11-02 15:42:31  classification/scale_features_or_not.svg      685   
2020-11-02 15:42:34                        timeseries/project      685   

                     cohort_id             ip    name  start_date    end_date  
date                                                                           
2020-07-13 14:40:31         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-07-13 23:23:10         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-07-14 08:30:16         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-07-14 10:37:01         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-07-14 11:09:36         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
...                        ...            ...     ...         ...         ...  
2020-11-02 15:20:36         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-11-02 15:20:50         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-11-02 15:42:31         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-11-02 15:42:31         59  99.76.233.212  Darden  2020-07-13  2021-01-12  
2020-11-02 15:42:34         59  99.76.233.212  Darden  2020-07-13  2021-01-12  

[2542 rows x 7 columns]

In [135]:
data_science[data_science.user_id == 485]

page_viewed  user_id  \
date                                                                     
2019-08-20 09:44:21                                         /      485   
2019-08-20 10:45:33                                         /      485   
2019-08-20 10:47:26                                         /      485   
2019-08-20 10:47:42  1-fundamentals/1.1-intro-to-data-science      485   
2019-08-20 10:47:42  1-fundamentals/modern-data-scientist.jpg      485   
...                                                       ...      ...   
2020-11-02 10:18:42      1-fundamentals/AI-ML-DL-timeline.jpg      485   
2020-11-02 10:19:02                    9-timeseries/2-acquire      485   
2020-11-02 10:19:35         9-timeseries/5.1-modeling-lesson1      485   
2020-11-02 11:57:36                                         /      485   
2020-11-02 11:57:55               timeseries/modeling-lesson1      485   

                     cohort_id              ip   name  start_date    end_date  
date                                                                           
2019-08-20 09:44:21         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:45:33         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:47:26         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:47:42         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:47:42         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
...                        ...             ...    ...         ...         ...  
2020-11-02 10:18:42         34  173.174.142.84  Bayes  2019-08-19  2020-01-30  
2020-11-02 10:19:02         34  173.174.142.84  Bayes  2019-08-19  2020-01-30  
2020-11-02 10:19:35         34  173.174.142.84  Bayes  2019-08-19  2020-01-30  
2020-11-02 11:57:36         34  173.174.142.84  Bayes  2019-08-19  2020-01-30  
2020-11-02 11:57:55         34  173.174.142.84  Bayes  2019-08-19  2020-01-30  

[2298 rows x 7 columns]

In [136]:
data_science[data_science.user_id == 475]

page_viewed  user_id  \
date                                                                     
2019-08-20 09:40:19                                         /      475   
2019-08-20 10:46:35  1-fundamentals/1.1-intro-to-data-science      475   
2019-08-20 10:46:36  1-fundamentals/modern-data-scientist.jpg      475   
2019-08-20 10:46:36      1-fundamentals/AI-ML-DL-timeline.jpg      475   
2019-08-20 10:50:11  1-fundamentals/1.1-intro-to-data-science      475   
...                                                       ...      ...   
2020-10-19 17:57:02                       sql/database-design      475   
2020-10-19 17:57:15                   python/intro-to-sklearn      475   
2020-10-19 17:57:36                  python/matplotlib-styles      475   
2020-10-19 17:57:56           classification/confusion-matrix      475   
2020-10-19 17:58:27           classification/confusion-matrix      475   

                     cohort_id              ip   name  start_date    end_date  
date                                                                           
2019-08-20 09:40:19         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:46:35         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:46:36         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:46:36         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:50:11         34    97.105.19.58  Bayes  2019-08-19  2020-01-30  
...                        ...             ...    ...         ...         ...  
2020-10-19 17:57:02         34  70.226.109.182  Bayes  2019-08-19  2020-01-30  
2020-10-19 17:57:15         34  70.226.109.182  Bayes  2019-08-19  2020-01-30  
2020-10-19 17:57:36         34  70.226.109.182  Bayes  2019-08-19  2020-01-30  
2020-10-19 17:57:56         34  70.226.109.182  Bayes  2019-08-19  2020-01-30  
2020-10-19 17:58:27         34  70.226.109.182  Bayes  2019-08-19  2020-01-30  

[2135 rows x 7 columns]

In [137]:
data_science[data_science.user_id == 476]

page_viewed  user_id  \
date                                                                     
2019-08-20 09:40:23                                         /      476   
2019-08-20 10:04:25                                         /      476   
2019-08-20 10:07:20                                         /      476   
2019-08-20 10:49:50  1-fundamentals/1.1-intro-to-data-science      476   
2019-08-20 10:49:50  1-fundamentals/modern-data-scientist.jpg      476   
...                                                       ...      ...   
2020-10-19 12:24:58                                 nlp/model      476   
2020-10-20 07:41:54                                 nlp/model      476   
2020-10-20 17:30:13                                 nlp/model      476   
2020-10-21 12:42:15                                         /      476   
2020-10-21 12:42:18                                 nlp/model      476   

                     cohort_id             ip   name  start_date    end_date  
date                                                                          
2019-08-20 09:40:23         34   97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:04:25         34   97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:07:20         34   97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:49:50         34   97.105.19.58  Bayes  2019-08-19  2020-01-30  
2019-08-20 10:49:50         34   97.105.19.58  Bayes  2019-08-19  2020-01-30  
...                        ...            ...    ...         ...         ...  
2020-10-19 12:24:58         34   24.54.121.33  Bayes  2019-08-19  2020-01-30  
2020-10-20 07:41:54         34   24.54.121.33  Bayes  2019-08-19  2020-01-30  
2020-10-20 17:30:13         34   24.54.121.33  Bayes  2019-08-19  2020-01-30  
2020-10-21 12:42:15         34  136.50.49.145  Bayes  2019-08-19  2020-01-30  
2020-10-21 12:42:18         34  136.50.49.145  Bayes  2019-08-19  2020-01-30  

[2046 rows x 7 columns]

In [138]:
web_dev.user_id.value_counts()

11     14870
64     14419
53     10874
1       6440
314     6419
       ...  
592        1
649        1
66         1
745        1
651        1
Name: user_id, Length: 649, dtype: int64

In [139]:
web_dev[web_dev.user_id == 11]

page_viewed  \
date                                                                     
2018-01-26 10:14:47                                                  /   
2018-01-26 10:14:51                                       javascript-i   
2018-01-26 10:15:02                             javascript-i/functions   
2018-01-26 11:46:29  appendix/further-reading/javascript/hoisting.html   
2018-01-26 11:46:39                                           appendix   
...                                                                ...   
2020-11-02 16:22:12                                        nlp/explore   
2020-11-02 16:42:02                         anomaly-detection/overview   
2020-11-02 16:42:02     anomaly-detection/AnomalyDetectionCartoon.jpeg   
2020-11-02 16:42:04        anomaly-detection/detecting-with-clustering   
2020-11-02 16:42:06   anomaly-detection/detecting-timeseries-anomalies   

                     user_id  cohort_id              ip    name  start_date  \
date                                                                          
2018-01-26 10:14:47       11          1    97.105.19.61  Arches  2014-02-04   
2018-01-26 10:14:51       11          1    97.105.19.61  Arches  2014-02-04   
2018-01-26 10:15:02       11          1    97.105.19.61  Arches  2014-02-04   
2018-01-26 11:46:29       11          1    97.105.19.61  Arches  2014-02-04   
2018-01-26 11:46:39       11          1    97.105.19.61  Arches  2014-02-04   
...                      ...        ...             ...     ...         ...   
2020-11-02 16:22:12       11         28  76.185.145.231   Staff  2014-02-04   
2020-11-02 16:42:02       11         28  76.185.145.231   Staff  2014-02-04   
2020-11-02 16:42:02       11         28  76.185.145.231   Staff  2014-02-04   
2020-11-02 16:42:04       11         28  76.185.145.231   Staff  2014-02-04   
2020-11-02 16:42:06       11         28  76.185.145.231   Staff  2014-02-04   

                       end_date  
date                             
2018-01-26 10:14:47  2014-04-22  
2018-01-26 10:14:51  2014-04-22  
2018-01-26 10:15:02  2014-04-22  
2018-01-26 11:46:29  2014-04-22  
2018-01-26 11:46:39  2014-04-22  
...                         ...  
2020-11-02 16:22:12  2014-02-04  
2020-11-02 16:42:02  2014-02-04  
2020-11-02 16:42:02  2014-02-04  
2020-11-02 16:42:04  2014-02-04  
2020-11-02 16:42:06  2014-02-04  

[14870 rows x 7 columns]

In [140]:
web_dev[web_dev.user_id == 64]

page_viewed  user_id  cohort_id  \
date                                                                 
2018-01-29 13:50:57                          /       64          1   
2018-01-29 13:51:00   mkdocs/search_index.json       64          1   
2018-01-29 13:51:04                mysql/users       64          1   
2018-01-30 10:13:09                          /       64          1   
2018-01-30 10:13:12                     spring       64          1   
...                                        ...      ...        ...   
2020-01-21 09:36:58                          /       64         52   
2020-01-21 09:37:00                      mysql       64         52   
2020-01-21 09:37:01                   java-iii       64         52   
2020-01-21 09:37:03                     spring       64         52   
2020-01-21 09:37:12  spring/fundamentals/views       64         52   

                                 ip    name  start_date    end_date  
date                                                                 
2018-01-29 13:50:57    97.105.19.61  Arches  2014-02-04  2014-04-22  
2018-01-29 13:51:00    97.105.19.61  Arches  2014-02-04  2014-04-22  
2018-01-29 13:51:04    97.105.19.61  Arches  2014-02-04  2014-04-22  
2018-01-30 10:13:09    97.105.19.61  Arches  2014-02-04  2014-04-22  
2018-01-30 10:13:12    97.105.19.61  Arches  2014-02-04  2014-04-22  
...                             ...     ...         ...         ...  
2020-01-21 09:36:58  12.106.208.194  Europa  2019-11-04  2020-04-17  
2020-01-21 09:37:00  12.106.208.194  Europa  2019-11-04  2020-04-17  
2020-01-21 09:37:01  12.106.208.194  Europa  2019-11-04  2020-04-17  
2020-01-21 09:37:03  12.106.208.194  Europa  2019-11-04  2020-04-17  
2020-01-21 09:37:12  12.106.208.194  Europa  2019-11-04  2020-04-17  

[14419 rows x 7 columns]

In [141]:
web_dev[web_dev.user_id == 53]

page_viewed  user_id  \
date                                                                  
2018-01-28 18:48:19                             index.html       53   
2018-01-28 18:48:24                               html-css       53   
2018-01-28 18:49:20                           javascript-i       53   
2018-01-28 18:49:23              javascript-i/conditionals       53   
2018-01-28 18:50:41              javascript-i/conditionals       53   
...                                                    ...      ...   
2020-11-02 09:09:53  examples/javascript/dom-query-js.html       53   
2020-11-02 15:27:42                        appendix/slides       53   
2020-11-02 16:09:40                             index.html       53   
2020-11-02 16:09:42                           javascript-i       53   
2020-11-02 16:09:45           javascript-i/bom-and-dom/dom       53   

                     cohort_id            ip    name  start_date    end_date  
date                                                                          
2018-01-28 18:48:19         14  67.10.161.55  Lassen  2016-07-18  2016-11-10  
2018-01-28 18:48:24         14  67.10.161.55  Lassen  2016-07-18  2016-11-10  
2018-01-28 18:49:20         14  67.10.161.55  Lassen  2016-07-18  2016-11-10  
2018-01-28 18:49:23         14  67.10.161.55  Lassen  2016-07-18  2016-11-10  
2018-01-28 18:50:41         14  67.10.161.55  Lassen  2016-07-18  2016-11-10  
...                        ...           ...     ...         ...         ...  
2020-11-02 09:09:53         28  67.10.174.56   Staff  2014-02-04  2014-02-04  
2020-11-02 15:27:42         28  67.10.174.56   Staff  2014-02-04  2014-02-04  
2020-11-02 16:09:40         28  67.10.174.56   Staff  2014-02-04  2014-02-04  
2020-11-02 16:09:42         28  67.10.174.56   Staff  2014-02-04  2014-02-04  
2020-11-02 16:09:45         28  67.10.174.56   Staff  2014-02-04  2014-02-04  

[10874 rows x 7 columns]

In [142]:
web_dev[web_dev.user_id == 1]

page_viewed  user_id  \
date                                                                          
2018-01-26 09:55:03                                              /        1   
2018-01-26 09:56:02                                        java-ii        1   
2018-01-26 09:56:05            java-ii/object-oriented-programming        1   
2018-01-26 09:56:06             slides/object_oriented_programming        1   
2018-01-26 10:40:15                         javascript-i/functions        1   
...                                                            ...      ...   
2020-10-30 09:33:22  appendix/extra-challenges/java-file-io-extras        1   
2020-11-02 11:13:31                                              /        1   
2020-11-02 11:13:34                        classification/overview        1   
2020-11-02 11:13:35       classification/scale_features_or_not.svg        1   
2020-11-02 11:13:40                    timeseries/modeling-lesson1        1   

                     cohort_id            ip     name  start_date    end_date  
date                                                                           
2018-01-26 09:55:03          8  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:02          8  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:05          8  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 09:56:06          8  97.105.19.61  Hampton  2015-09-22  2016-02-06  
2018-01-26 10:40:15          8  97.105.19.61  Hampton  2015-09-22  2016-02-06  
...                        ...           ...      ...         ...         ...  
2020-10-30 09:33:22         28  66.69.40.223    Staff  2014-02-04  2014-02-04  
2020-11-02 11:13:31         28  66.69.40.223    Staff  2014-02-04  2014-02-04  
2020-11-02 11:13:34         28  66.69.40.223    Staff  2014-02-04  2014-02-04  
2020-11-02 11:13:35         28  66.69.40.223    Staff  2014-02-04  2014-02-04  
2020-11-02 11:13:40         28  66.69.40.223    Staff  2014-02-04  2014-02-04  

[6440 rows x 7 columns]

In [143]:
import requests

In [144]:
# query = "97.105.19.61"
url = 'http://ip-api.com/csv/97.105.19.61'
response = requests.get(url)
var = response.text

In [145]:
var


'success,United States,US,TX,Texas,Dallas,75243,32.9103,-96.7283,America/Chicago,Spectrum,Codeup LLC,AS11427 Charter Communications Inc,97.105.19.61\n'

In [154]:
for index, row in darden.iterrows():
    location = row['ip']
    url = 'http://ip-api.com/csv/' + location
    response = requests.get(url)
    var = response.text
    print(var)

success,United States,US,TX,Texas,Austin,78723,30.3023,-97.6914,America/Chicago,"AT&T Services, Inc.",AT&T Corp,"AS7018 AT&T Services, Inc.",76.201.20.193

success,United States,US,TX,Texas,San Antonio,78251,29.4551,-98.6498,America/Chicago,Charter Communications,Spectrum,AS11427 Charter Communications Inc,24.28.146.155

success,United States,US,TX,Texas,San Antonio,78250,29.5,-98.6744,America/Chicago,Google Fiber Inc.,Google Fiber Inc,AS16591 Google Fiber Inc.,136.50.56.155

success,United States,US,TX,Texas,San Antonio,78251,29.4551,-98.6498,America/Chicago,Charter Communications,Spectrum,AS11427 Charter Communications Inc,24.28.146.155

success,United States,US,TX,Texas,San Antonio,78251,29.4551,-98.6498,America/Chicago,Charter Communications,Spectrum,AS11427 Charter Communications Inc,24.28.146.155

success,United States,US,TX,Texas,San Antonio,78251,29.4551,-98.6498,America/Chicago,Charter Communications,Spectrum,AS11427 Charter Communications Inc,24.28.146.155

success,United Sta

ReadTimeout: HTTPConnectionPool(host='ip-api.com', port=80): Read timed out. (read timeout=None)

In [152]:
darden.head()

page_viewed  user_id  \
date                                                                     
2020-07-13 14:34:44                                         /      678   
2020-07-13 14:37:22                                         /      679   
2020-07-13 14:38:28                                         /      680   
2020-07-13 14:39:21            13-advanced-topics/1-tidy-data      679   
2020-07-13 14:39:36  1-fundamentals/1.1-intro-to-data-science      679   

                     cohort_id             ip    name  start_date    end_date  
date                                                                           
2020-07-13 14:34:44         59  76.201.20.193  Darden  2020-07-13  2021-01-12  
2020-07-13 14:37:22         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:38:28         59  136.50.56.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:21         59  24.28.146.155  Darden  2020-07-13  2021-01-12  
2020-07-13 14:39:36         59  24.28.146.155  Darden  2020-07-13  2021-01-12